### PR TITLE
Warning to users who try to login without course completion that they must complete course

### PR DESCRIPTION
**Pull request by dtgreiner**
## _Merging this PR_
- use the squash-merge strategy for PRs targeting a release-X branch
- use a merge-commit or rebase strategy for PRs targeting the stable branch

## Description
[//]: # (Summarize changes and include links related issue. List any new dependencies)
Add a new flag for the TalentLMS configs allowing/disallowing the automated creation of accounts in the TalentLMS domain.

Create a captive portal for users that require training but we are unable to automate a new account creation for them in at least one TalentLMS subdomain. This portal includes links to the TalentLMS domains where accounts creation needs to be done and informs them to contact an admininstrator to setup and link these accounts.

## Type of change
[//]: # (e.g., Bug fix, New feature, Code clean-up, Dependency update)
New feature

## Checklist before requesting review
[//]: # (Remove any items that are not applicable)
- [X] I have performed a self-review of my code
- [X] I have run the code that is being changed under ideal conditions, and it doesn't fail
- [ ] If adding a new endpoint / exposing data in a new way, I have:
  - [ ] ensured the API can't leak data from other data sources
  - [ ] ensured this does not introduce N+1s
  - [ ] ensured permissions and visibility checks are performed in the right places
- [ ] I have updated the documentation (or not applicable)
- [ ] I have added spec tests (or not applicable)
- [ ] I have provided testing instructions in this PR or the related issue (or not applicable)

### CODE REVIEW

**Comment by dtgreiner** [comment]
Adding screenshots of the new/updated pages here:

Screenshot 2024-10-31 at 3 52 54 PM
Screenshot 2024-10-31 at 4 07 20 PM
Screenshot 2024-10-31 at 4 07 51 PM

File: app/views/user_training/required_trainings.haml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        %thead
+          %tr
+            %th Course Name
+            %th Course ID
+        %tbody
+          - required_courses_for_config = (@courses & config.courses)
+          - required_courses_for_config.each do |course|
+            %tr
+              %td= course.name
+              %td= course.courseid
```
Comment (eanders) [inline app/views/user_training/required_trainings.haml]: Can you add a "Completed" column to this with a yes/no check/x for completed/incomplete?  That way people can see the progress toward being able to access the site.

**Review by eanders — APPROVED** [review]
@dtgreiner this looks great, just one request inline.

File: app/views/user_training/required_trainings.haml (hunk 1 above)
Comment (dtgreiner) [reply app/views/user_training/required_trainings.haml]: Updated the portal to display this info in [this commit](https://github.com/greenriver/hmis-warehouse/pull/4887/commits/a820796caba836fcd7d38d854f5958b46021b677)

File: app/models/talentlms/course.rb [hunk 2]
```diff
@@ -60,5 +60,13 @@ def self.remove_course(course_id)
     def self.collection_for_form
       Talentlms::Course.all.map { |c| [c.name_with_subdomain, c.id] }
     end
+
+    # Check local data for course completion by user
+    def completed_by?(user)
+      login = Talentlms::Login.find_by(config: config, user: user)
+      return unless login # Login does not exist, user has not completed training
+
+      Talentlms::CompletedTraining.all.detect { |ct| ct.login_id == login.id && ct.course_id == id }.present?
```
Comment (eanders) [inline app/models/talentlms/course.rb]: This looks like it is bringing back all the CompletedTraining records to determine if any match the login and course.  Couldn't it just:
```suggestion
      Talentlms::CompletedTraining.where(login_id: login.id, course_id: id).exists?
```

File: app/views/user_training/required_trainings.haml [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+            %th Course Completed
+        %tbody
+          - required_courses_for_config = (@courses & config.courses)
+          - required_courses_for_config.each do |course|
+            %tr
+              %td= course.name
+              %td= course.courseid
+              %td
+                - if course.completed_by?(current_user)
+                  %i.icon-checkmark
```
Comment (eanders) [inline app/views/user_training/required_trainings.haml]:
```suggestion
              %td= checkmark(course.completed_by?(current_user))
```

Should do the same thing.

File: app/models/talentlms/course.rb (hunk 2 above)
Comment (dtgreiner) [reply app/models/talentlms/course.rb]: yes, that's much better, adding it in.

**Pull request merged**